mindmeld/components/entity_resolver.py

```diff
@@ -51,51 +65,189 @@ class EntityResolver:
-        self._resource_loader = resource_loader
-        self._normalizer = resource_loader.query_factory.normalize
+        er_config = (
+            kwargs.pop("er_config", None) or
+            get_classifier_config("entity_resolution", app_path=app_path)
+        )
+        name = er_config.get("model_type", None)
+        cls.validate_resolver_name(name)
+        return ENTITY_RESOLVER_MODEL_MAPPINGS.get(name)(
+            app_path, resource_loader, entity_type, er_config, **kwargs
+        )
+
+    def fit(self):
+        raise NotImplementedError
+
+    def load(self):
+        raise NotImplementedError
+
+    def predict(self, entity):
+        raise NotImplementedError
+
+
+class EntityResolverBase(ABC):
+    """
+    Base class for Entity Resolvers
+    """
+
+    def __init__(self, app_path, resource_loader, entity_type, er_config, **kwargs):
+        """Initializes an entity resolver"""
+        self.app_path = app_path
+        self.resource_loader = resource_loader
         self.type = entity_type
+        self.er_config = er_config
+        self.kwargs = kwargs
+
+        self._app_namespace = get_app_namespace(self.app_path)
         self._is_system_entity = Entity.is_system_entity(self.type)
-        self._exact_match_mapping = None
-        self._er_config = get_classifier_config("entity_resolution", app_path=app_path)
-        self._es_host = es_host
-        self._es_config = {"client": es_client, "pid": os.getpid()}
-        self.ready = False
+        self.name = self.er_config.get("model_type")
+        self.dirty = False  # bool, True if exists any unsaved generated data that can be saved
+        self.ready = False  # bool, True if the model is fit by calling .fit()
 
         if self._is_system_entity:
             canonical_entities = []
         else:
-            canonical_entities = self._resource_loader.get_entity_map(self.type).get(
+            canonical_entities = self.resource_loader.get_entity_map(self.type).get(
                 "entities", []
             )
         self._no_canonical_entity_map = len(canonical_entities) == 0
 
-    @property
-    def _es_index_name(self):
-        return EntityResolver.ES_SYNONYM_INDEX_PREFIX + "_" + self.type
+        if self._use_double_metaphone:
+            self._invoke_double_metaphone_usage()
 
     @property
-    def _use_text_rel(self):
-        return self._er_config["model_type"] == "text_relevance"
+    def _use_double_metaphone(self):
+        return "double_metaphone" in self.er_config.get("phonetic_match_types", [])
+
+    def _invoke_double_metaphone_usage(self):
+        """
+        By default, resolvers are assumed to not support double metaphone usage
+        If supported, override this method definition in the derived class
+        (eg. see EntityResolverUsingElasticSearch)
+        """
+        logger.warning(
+            "%r not configured to use double_metaphone",
+            self.name
+        )
+        raise NotImplementedError
```

Comment:
I am not sure why the base class has to offer this method as an abstract method. I think a derived class should construct and call this function.

Reply:
Also, I dont see any class implementing this function

Reply:
This is (theoretically) not an abstract method. By default, we consider an entity resolver does not use phonetic match. If accidentally, in config, a phonetic type is provided when using a non- elastic-search resolver, it would raise an error. I kept the functionality intact for all derived classes (by adding it to base class) because in future, BERT based resolvers could also consider using double metaphone through some way.

(To second comment) Elastic Search resolver implements and overrides the base class' function. But other resokvers do not.